apps/utils/locust_tests/common/stats/stats_compare.py

```diff
@@ -1,212 +1,332 @@
 from common.stats.stats_loaders import StatsLoader
 
-TaskStatsOrPercentiles = Union[TaskStats, ResponseTimePercentiles]
-"""Type indicating a value that is either a TaskStats instance or a response time
-percentiles dictionary"""
 
-DEFAULT_DEVIANCE_FAILURE_THRESHOLD = 500.0
-"""The default percent threshold that stats should not exceed or be worse than compared to a previous or average run"""
+class StatBetterIf(str, Enum):
+    """Enum representing whether a given stat is better if its value is smaller or larger"""
+
+    SMALLER = "SMALLER"
+    LARGER = "LARGER"
+
+
+class StatCompareResult(str, Enum):
+    """Enum representing the result of a particular stat comparison"""
+
+    PASS = "PASS"
+    WARNING = "WARNING"
+    FAILURE = "FAILURE"
+
+
+@dataclass
+class StatsComparisonMetadata:
+    """Dataclass representing how a given stat should be compared against a baseline value of the
+    same stat"""
+
+    name: str
+    """Name of the stat"""
+    better_if: StatBetterIf
+    """Enum indicating whether the stat is "better" if its value is smaller or larger in comparison
+    to a baseline. I.e. number of requests is better if larger, but response time is better if
+    smaller"""
+    warn_percent: float
+    """A percent threshold that if the percent ratio between the current stat and a baseline exceeds
+    will be reported as a warning. Does not cause a failure for the run -- only used to indicate
+    anomalously high stats"""
+    fail_percent: float
+    """A percent threshold that if the percent ratio between the current stat and a baseline exceeds
+    will be reported as a failure. Causes a failure for the run"""
 
 
 @dataclass
-class StatCompareResult:
+class AllComparisonMetadata:
+    """Dataclass representing comparison metadata defining how stats are compared for both
+    aggregated totals stats and individual tasks' stats"""
+
+    totals: List[StatsComparisonMetadata]
+    """List of metadata about each stat for aggregated totals"""
+    tasks: List[StatsComparisonMetadata]
+    """List of metadata about each stat for each invidividual task"""
+
+
+@dataclass
+class StatComparison:
     """Dataclass representing the result of a comparison between the baseline/previous value of a
     TaskStats stat and the current value of said stat"""
 
     stat: str
     """The name of the stat"""
     percent: float
-    """The percent value of this stat's value versus a previous run's, i.e. this stat is percent% times worse than
-    the previous run or average of all runs. A value of 100 means no change (i.e. current stat is 100%
-    of a previous snapshot of the same stat), whereas values lower than 100 generally indicate
-    positive change and values greater than 100 indicate negative change"""
+    """The percent value of this stat's value versus a previous run's, i.e. this stat is percent%
+	times worse than the previous run or average of all runs. A value of 100 means no change (i.e.
```

Comment:
Just curous....would a negative val=ue (if supported) represent an improvement?

Reply:
When I wrote this I didn't actually realize that what `percent` is is actually the "percent ratio" between the current stat and previous stat. So, you can think of it as:

$$\frac{current}{previous}\times 100 = percent$$
or
$$\frac{previous}{current}\times 100 = percent$$
(depending on whether or not the stat's value is "better" for smaller values [i.e. response times] or for larger values [number of requests])

Since it's a ratio, anything below 100% means that the current stat's value is _better_ than the baseline's value. As in, if the current `total_reqs_per_second` is 100 and the baseline's value was 95:

$$\frac{previous}{current}\times 100 = \frac{95}{100}\times 100 = 95\\%$$

So percent ratios under 100% indicate improvement, a percent ratio of _exactly_ 100% indicates no change, and a percent ratio greater than 100% indicates negative change. It shouldn't be possible for `percent` be negative, since it's really just a ratio between the two stats.